src/components/lab/canvas/molecules/CanvasMethane.js

```diff
@@ -7,85 +8,142 @@ import {
+
+  // variables for determining center points of atoms in this molecule
+  // we know that after INTERVALS_TO_REACH_MOLECULE_CENTER, the radiation lines have reached the center of the molecule
+  // at this point, if the spectrum is INFRARED, the molecule should begin oscillating
   const hydrogenAtomRadius = CANVAS_ATOM_DIMENSIONS[HYDROGEN.size];
   const carbonAtomRadius = CANVAS_ATOM_DIMENSIONS[CARBON.size];
-  const carbonAtomInitialCenterPoint = {
-    x,
+  const carbonAtomCenterPoint = {
+    x:
+      intervalCount > INTERVALS_TO_REACH_MOLECULE_CENTER &&
+      spectrum === SPECTRUMS.INFRARED
+        ? x +
+          CANVAS_METHANE_OSCILLATION_AMPLITUDES.CARBON_ATOM *
+            Math.sin(
+              intervalCount * Y_SHIFT_PER_INTERVAL * INFRARED_RADIATION_PERIOD,
+            )
+        : x,
     y,
   };
-  const topLeftHydrogenAtomInitialCenterPoint = {
-    x: x - CANVAS_METHANE_FIRST_ATOM_X_ADJUSTMENT_FACTOR,
-    y:
-      y -
-      carbonAtomRadius -
-      hydrogenAtomRadius +
-      CANVAS_METHANE_FIRST_ATOM_Y_ADJUSTMENT_FACTOR,
+  const topLeftHydrogenAtomCenterPoint = {
+    x:
+      intervalCount > INTERVALS_TO_REACH_MOLECULE_CENTER &&
+      spectrum === SPECTRUMS.INFRARED
+        ? x -
+          CANVAS_METHANE_TOP_LEFT_HYDROGEN_ATOM_X_ADJUSTMENT_FACTOR +
+          CANVAS_METHANE_OSCILLATION_AMPLITUDES.TOP_LEFT_HYDROGEN_ATOM *
+            Math.sin(
+              intervalCount * Y_SHIFT_PER_INTERVAL * INFRARED_RADIATION_PERIOD,
+            )
+        : x - CANVAS_METHANE_TOP_LEFT_HYDROGEN_ATOM_X_ADJUSTMENT_FACTOR,
+    y: y - carbonAtomRadius - hydrogenAtomRadius,
   };
-  const topRightHydrogenAtomInitialCenterPoint = {
-    x: x + CANVAS_METHANE_SECOND_ATOM_X_ADJUSTMENT_FACTOR,
+  const topRightHydrogenAtomCenterPoint = {
+    x: x + CANVAS_METHANE_TOP_RIGHT_HYDROGEN_ATOM_X_ADJUSTMENT_FACTOR,
     y:
-      y -
-      carbonAtomRadius -
-      hydrogenAtomRadius +
-      CANVAS_METHANE_SECOND_ATOM_Y_ADJUSTMENT_FACTOR,
+      intervalCount > INTERVALS_TO_REACH_MOLECULE_CENTER &&
+      spectrum === SPECTRUMS.INFRARED
+        ? y -
+          carbonAtomRadius -
+          0.25 * hydrogenAtomRadius +
+          CANVAS_METHANE_OSCILLATION_AMPLITUDES.TOP_RIGHT_HYDROGEN_ATOM *
+            Math.sin(
+              intervalCount * Y_SHIFT_PER_INTERVAL * INFRARED_RADIATION_PERIOD,
+            )
+        : y - carbonAtomRadius - 0.25 * hydrogenAtomRadius,
   };
-  const bottomRightHydrogenAtomInitialCenterPoint = {
-    x: x + CANVAS_METHANE_THIRD_ATOM_X_ADJUSTMENT_FACTOR,
+  const bottomRightHydrogenAtomCenterPoint = {
+    x: x + CANVAS_METHANE_BOTTOM_RIGHT_HYDROGEN_ATOM_X_ADJUSTMENT_FACTOR,
     y:
-      y +
-      carbonAtomRadius +
-      hydrogenAtomRadius -
-      CANVAS_METHANE_THIRD_ATOM_Y_ADJUSTMENT_FACTOR,
+      intervalCount > INTERVALS_TO_REACH_MOLECULE_CENTER &&
+      spectrum === SPECTRUMS.INFRARED
+        ? y +
+          carbonAtomRadius +
+          0.25 * hydrogenAtomRadius -
```

Comment:
Can we factor out this 0.25? Not sure what it's doing.

Reply:
You might want to refactor to a function in order to avoid duplicate code (just changing the +/- sign).

Reply:
for the 0.25... it's just saying: this atom's y position is: the position of the center of molecule (y), plus one carbon radius, plus one-quarter of a hydrogen atom radius. I could come up with a constant (`CANVAS_METHANE_BOTTOM_RIGHT_HYDROGEN_ATOM_HYDROGEN_ATOM_RADIUS_ADJUSTMENT_FACTOR`) and add it to `constants.js`, but... it feels too much. what do you think? I felt it's clear that it's just adjusting the positioning. I could write it as `(1 / 4) * hydrogenAtomRadius` if it's clearer?

Reply:
you're right about duplicate code in these components in general. I made a fix which reduced duplication significantly, by moving certain variables (`shouldOscillate` and `oscillationFormula`) to the parent component and passing them as props. otherwise, we have certain repetitions like: `x - CANVAS_METHANE_TOP_LEFT_HYDROGEN_ATOM_X_ADJUSTMENT_FACTOR`. Maybe it's worth creating another object: `topLeftHydrogenAtomInitialCenterPoints: {x: , y:}`, and then referencing that, but I feel this might just make the code more cumbersome. I also considered renaming the `x` and `y` received as props to `moleculeCenterX` and `moleculeCenterY`, for additional clarity, but in the end felt that already variable names are so long it might not be worth it. if you have ideas for how to re-organize the code in `CanvasMethane`, `CanvasCarbonDioxide`, and the other molecules, would love to hear them!